File: cliente-python/src/noise/bit_noise.py
from random import random


def apply_bit_flip_noise(codeword_bits: str, bit_flip_probability: float) -> str:
    if not 0.0 <= bit_flip_probability <= 1.0:
        raise ValueError("La probabilidad de flip debe estar entre 0.0 y 1.0.")

    noisy_bits = []

    for bit in codeword_bits:
        if random() < bit_flip_probability:
            noisy_bits.append(flip_bit(bit))
        else:
            noisy_bits.append(bit)

    return "".join(noisy_bits)


def flip_bit(bit: str) -> str:
    return "1" if bit == "0" else "0"

File: servidor-java/src/link/BitNoise.java
package link;

import java.util.concurrent.ThreadLocalRandom;


public final class BitNoise {
    private BitNoise() {
    }

    public static String apply(String protectedBits, double bitFlipProbability) {
        if (bitFlipProbability < 0 || bitFlipProbability > 1) {
            throw new IllegalArgumentException("La probabilidad debe estar entre 0.0 y 1.0.");
        }

        StringBuilder noisyBits = new StringBuilder();

        for (int index = 0; index < protectedBits.length(); index++) {
            char originalBit = protectedBits.charAt(index);
            boolean shouldFlip = ThreadLocalRandom.current().nextDouble() < bitFlipProbability;
            noisyBits.append(shouldFlip ? flipBit(originalBit) : originalBit);
        }

        return noisyBits.toString();
    }

    private static char flipBit(char bit) {
        return bit == '0' ? '1' : '0';
    }
}
